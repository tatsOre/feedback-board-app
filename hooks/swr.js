import { isArray, isEmpty, isEqual } from 'lodash'
import { useCallback, useEffect, useMemo, useState } from 'react'
import useSWR, { ConfigInterface } from 'swr'

/*
 * Compare two objects by reducing an array of keys in obj1, having the
 * keys in obj2 as the initial value of the result. Key points:
 *
 * — All keys of obj2 are initially in the result.
 *
 * — If the loop finds a key (from obj1, remember) not in obj2, it adds
 * it to the result.
 *
 * — If the loop finds a key that are both in obj1 and obj2, it compares
 * the value. If it’s the same value, the key is removed from the result.
 */
export function getObjectDifference(obj1, obj2) {
  const diff = Object.keys(obj1).reduce((result, key) => {
    if (!obj2.hasOwnProperty(key)) {
      result.push(key)
    }
    return result
  }, Object.keys(obj2))
  return Object.fromEntries(
    diff.map((key) => {
      return [key, obj2[key]]
    })
  )
}

export function useCrud(url, key, fetchOptions) {
  const [loading, setIsLoading] = useState(true)

  const loadingTimeout = () => {
    setIsLoading(false)
  }

  const fetch = useCallback(async (url) => {
    const response = await fetch(url)
    return response
  }, [])

  const { data, error, isValidating, mutate } = useSWR(url, fetch, {
    ...fetchOptions,
  })

  useEffect(() => {
    if (isValidating) {
      setIsLoading(true)
      return
    }
    setTimeout(loadingTimeout, 500)
  }, [isValidating])

  const create = useCallback(
    async (newObject, shouldRevalidate = false) => {
      const response = await fetch(url, {
        body: newObject,
        method: 'POST',
      })
      const result = response
      if (data && mutate) {
        let newData = data
        if (isArray(data)) {
          newData = data.concat(result)
        }
        await mutate([...new Set(newData)], shouldRevalidate)
      }
      return result
    },
    [url, data, mutate]
  )
  
  const createMultiple = useCallback(
    async (newObjects, shouldRevalidate = false) => {
      const response = await fetch(url, {
        body: newObjects,
        method: 'POST',
      })
      const result = response
      if (data && mutate) {
        await mutate([...data, ...result], shouldRevalidate)
      }
      return result
    },
    [url, data, mutate]
  )
  const remove = useCallback(
    async (body, shouldRevalidate = false) => {
      const response = await fetch(url, {
        body,
        method: 'DELETE',
      })
      const result = response
      if (data && mutate) {
        if (isArray(result)) {
          const updatedObjects = [...data].filter((current) => {
            const isDeleted = result.find(
              (result) => result[key] === current[key]
            )
            return !isDeleted
          })
          await mutate(
            result.length === 0 ? [] : updatedObjects,
            shouldRevalidate
          )
        } else {
          const deletedIndex = data.findIndex(
            (object) => object[key] === result[key]
          )
          if (deletedIndex >= 0) {
            const updatedObjects = [...data]
            updatedObjects.splice(deletedIndex, 1)
            await mutate(updatedObjects, shouldRevalidate)
          }
        }
      }
      return result
    },
    [url, data, key, mutate]
  )
  const removeMultiple = useCallback(
    async (ids, shouldRevalidate = false) => {
      const response = await fetch(url, {
        body: ids,
        method: 'DELETE',
      })
      const results = response
      if (data && mutate) {
        const updatedObjects = [...data].filter((current) => {
          const isDeleted = results.find(
            (result) => result[key] === current[key]
          )
          return !isDeleted
        })
        await mutate(updatedObjects, shouldRevalidate)
        return results
      }
    },
    [url, data, key, mutate]
  )
  const update = useCallback(
    async (updatedObject, shouldRevalidate = false) => {
      const currentObjectIndex = data.findIndex(
        (object) => object[key] === updatedObject[key]
      )
      const currentObject = data[currentObjectIndex]
      const diff = currentObject
        ? getObjectDifference(currentObject, updatedObject)
        : null
      if (!diff) {
        throw new Error('Update Failed')
      }
      if (isEmpty(diff)) {
        return currentObject
      }
      const response = await fetch(url, {
        body: { ...diff, id: updatedObject[key] },
        method: 'PATCH',
      })
      if (data && mutate) {
        const updatedObjects = [...data]
        updatedObjects.splice(currentObjectIndex, 1, response)
        await mutate(updatedObjects, shouldRevalidate)
      }
      return response
    },
    [url, data, mutate, key]
  )
  const updateMultiple = useCallback(
    async (updatedObjects, shouldRevalidate = false) => {
      const currentObjects = data.filter((object) =>
        updatedObjects.find((updated) => object[key] === updated[key])
      )
      if (!currentObjects || currentObjects <= 0) {
        throw new Error('Update Failed')
      }
      const diffs = currentObjects.map((currentObject) => {
        const updatedObject = updatedObjects.find(
          (updated) => updated[key] === currentObject[key]
        )
        return {
          ...getObjectDifference(currentObject, updatedObject),
          id: updatedObject[key],
        }
      })
      if (diffs.length <= 0) {
        return currentObjects
      }
      const response = await fetch(url, {
        body: { ...diffs },
        method: 'PATCH',
      })
      if (data && mutate) {
        const updatedObjects = [...data].map((current) => {
          if (current[key] === response[key]) {
            return response
          }
          return current
        })
        await mutate(updatedObjects, shouldRevalidate)
      }
      return response
    },
    [url, data, mutate, key]
  )
  const memoizedData = useMemo(
    () => (!isEmpty(data) ? filterDeleted(data) : []),
    [data]
  )
  return {
    create,
    createMultiple,
    fetch: { data: memoizedData, error, loading, mutate },
    remove,
    removeMultiple,
    update,
    updateMultiple,
  }
}
