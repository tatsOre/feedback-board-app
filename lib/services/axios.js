import { Axios } from 'axios'

const axios = new Axios({
  baseURL: '/api',
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  },
})

const AxiosAPIService = {
  getAll: async () => {
    const response = await axios.get('/feedbacks')
    return JSON.parse(response.data).data
  },

  post: async (data, endpoint = '/feedbacks') => {
    const response = await axios.post(endpoint, JSON.stringify(data))
    const doc = JSON.parse(response.data)
    if (!doc.success) throw new Error(response.statusText)
    return doc.data
  },

  get: async (endpoint) => {
    const response = await axios.get(endpoint)
    const doc = JSON.parse(response.data)
    if (!doc.success) throw new Error(response.statusText)
    return doc.data
  },

  update: async (id, data) => {
    const response = await axios.put(`/feedbacks/${id}`, JSON.stringify(data))
    const doc = JSON.parse(response.data)
    if (!doc.success) throw new Error(response.statusText)
    return doc.data
  },

  delete: async (id, uid) => {
    const response = await axios.delete(`/feedbacks/${id}`, {
      data: JSON.stringify({ user: uid }),
    })
    const doc = JSON.parse(response.data)
    if (!doc.success) throw new Error(response.statusText)
    return doc.data
  },
}

export { AxiosAPIService }
