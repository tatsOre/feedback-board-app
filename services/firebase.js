import admin from '../firebase/nodeApp'

export async function getAllFeedbacks() {
  const db = admin.firestore()
  const result = await db.collection('feedbacks').get()
  const docs = result.docs.map((fd) => ({
    id: fd.id,
    ...fd.data(),
  }))
  const schema = {
    'in-progress': [],
    live: [],
    planned: [],
    suggestion: [],
  }
  return docs.reduce((acc, req) => {
    let status = req.status
    acc[status] = acc[status].length ? [...acc[status], req] : [req]
    return acc
  }, schema)
}

export async function getFeedbackById(id) {
  const db = admin.firestore()
  const doc = await db.collection('feedbacks').doc(id).get()
  if (doc.exists) {
    return doc.data()
  }
  throw new Error('The document does not exist')
}
