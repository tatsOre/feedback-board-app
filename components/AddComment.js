import { useState } from 'react'
import { mutate } from 'swr'
import { AxiosAPIService } from '../lib/services/axios'
import Button from './Buttons/Default'
import ErrorMessage from './Error/DefaultError'

export default function AddComment({ data, user }) {
  const [{ content, error }, setComment] = useState({
    content: '',
    error: '',
  })

  const charsLeft = 250 - content.length

  const onChange = ({ target }) =>
    setComment({ content: target.value, error: '' })

  const onSubmit = async (event) => {
    event.preventDefault()
    if (!content)
      return setComment((state) => ({ ...state, error: "Can't be empty!" }))

    try {
      const comments = [...data.comments, { content, user: user._id }]

      const doc = await AxiosAPIService.update(data._id, { comments })

      mutate(`/feedbacks/slug?q=${data.slug}`, doc, false)

      setComment({ content: '', error: '' })
    } catch (error) {
      setComment((state) => ({ ...state, error: error.message }))
    }
  }

  return (
    <form
      onSubmit={onSubmit}
      className="bg-white rounded-10 p-6 md:px-8 mt-6 flex flex-wrap justify-between items-center"
    >
      <h2 className="w-full text-indigo-800 text-lg mb-6">Add Comment</h2>
      <textarea
        value={content}
        onChange={onChange}
        className={`w-full text-[13px] md:text-[15px] text-indigo-800 bg-indigo-100 rounded-5 p-4 mb-2 ${
          error ? 'border-red-900' : 'border-indigo-100 hover:border-blue-900'
        } border cursor-pointer`}
        placeholder="Type your comment here"
        rows="3"
        maxLength="250"
      />
      {error && <ErrorMessage text={error} style={{ width: '100%' }} />}

      <p className="text-indigo-500 text-[13px]">
        {charsLeft} character{charsLeft > 1 || !charsLeft ? 's' : ''} left
      </p>
      <Button type="submit">Post Comment</Button>
    </form>
  )
}
