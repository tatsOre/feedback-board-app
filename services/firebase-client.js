import { createFirebaseApp } from '../firebase/clientApp'
import {
  addDoc,
  arrayRemove,
  arrayUnion,
  collection,
  deleteDoc,
  doc,
  getDoc,
  getDocs,
  getFirestore,
  query,
  updateDoc,
  where,
} from 'firebase/firestore'

/**
 * Create a new document in the 'feedbacks' collection
 * @param   {Object}  data  Form input values
 * @returns {Object}
 *    Error:   if the document already exists | values are missing
 *    Success: id and slug of the newly document
 */
export async function createFeedback(data) {
  if (!data.title || !data.description) {
    throw new Error('Missing title or description')
  }

  const app = createFirebaseApp()
  const db = getFirestore(app)
  const q = query(collection(db, 'feedbacks'), where('slug', '==', data.slug))
  const snapshot = await getDocs(q)

  if (!snapshot.empty) {
    throw new Error('The document already exist')
  }
  const docRef = await addDoc(collection(db, 'feedbacks'), {
    ...data,
    status: 'suggestion',
    upvotes: 0,
    comments: [],
    created: new Date().toISOString(),
  }).catch((error) => {
    throw new Error(error.message)
  })
  return { id: docRef.id, slug: data.slug }
}

/**
 * Update a feedback document
 * @param   {Object}  data  Newly provided data
 * @returns {Object}
 *    Error:   if the document does not exist | values are missing | Firestore error
 *    Success: nothing
 */
export async function updateFeedback(data) {
  if (!data.title || !data.description) {
    throw new Error('Missing title or description')
  }

  const app = createFirebaseApp()
  const db = getFirestore(app)
  const docRef = doc(db, 'feedbacks', data.id)
  const docSnap = await getDoc(docRef)

  if (!docSnap.exists()) {
    throw new Error('The document does not exist')
  }
  await updateDoc(docRef, {
    ...data,
    updated: new Date().toISOString(),
  }).catch((error) => {
    throw new Error(error.message)
  })
}

/**
 * Get document by field
 * @param   {String}  field  Could be slug or title
 * @param   {String}  value  Value to look for
 * @returns {Object}
 *    Error:   if the document does not exist
 *    Success: feedback document with Firestore ID
 */
export async function getFeedbackByField(field, value) {
  const app = createFirebaseApp()
  const db = getFirestore(app)
  const q = query(collection(db, 'feedbacks'), where(field, '==', value))
  const snapshot = await getDocs(q)

  if (snapshot.empty) {
    throw new Error('The document does not exist')
  }
  const docs = snapshot.docs.map((doc) => ({ ...doc.data(), id: doc.id }))
  return docs[0]
}

/**
 * Get document by id
 * @param   {String}  id  Feedback Document ID
 * @returns {Object}
 *    Error:   if the document does not exist
 *    Success: Firestore Document
 */
export async function getFeedbackById(fdid) {
  const app = createFirebaseApp()
  const db = getFirestore(app)
  const docRef = doc(db, 'feedbacks', fdid)
  const docSnap = await getDoc(docRef)

  if (!docSnap.exists()) {
    throw new Error('The document does not exist')
  }
  return { ...docSnap.data(), id: docSnap.id }
}

export async function updateFeedbackComments(data, content, user) {
  const app = createFirebaseApp()
  const db = getFirestore(app)
  const fdRef = doc(db, 'feedbacks', data.id)

  const comment = {
    id: data.comments.length + 1,
    content,
    user: {
      image: user.image,
      name: user.name,
      username: user.username,
    },
  }

  let updatedDoc = null
  try {
    await updateDoc(fdRef, {
      updated: new Date().toISOString(),
      comments: arrayUnion(comment),
    })
    updatedDoc = await getFeedbackById(data.id)
  } catch (error) {
    console.log(error)
  }
  return updatedDoc
}

export async function updateFeedbackReplies(data, reply, cmid) {
  const app = createFirebaseApp()
  const db = getFirestore(app)
  const fdRef = doc(db, 'feedbacks', data.id)

  const updatedComments = data.comments.map((c) => {
    if (c.id === cmid) {
      return { ...c, replies: c.replies ? [...c.replies, reply] : [reply] }
    }
    return c
  })

  await updateDoc(fdRef, {
    updated: new Date().toISOString(),
    comments: updatedComments,
  }).catch((error) => {
    throw new Error(error.message)
  })
  return updatedComments
}

export async function deleteFeedback(fdid, user) {
  const app = createFirebaseApp()
  const db = getFirestore(app)

  const docRef = doc(db, 'feedbacks', fdid)
  await deleteDoc(docRef).catch((error) => {
    throw new Error(error.message)
  })
  await updateDoc(doc(db, 'users', user.id), {
    upvoted: arrayRemove(fdid),
  }).catch((error) => {
    throw new Error(error.message)
  })
}
