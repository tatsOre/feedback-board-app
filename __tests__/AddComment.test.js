import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import AddComment from '../components/AddComment'

// Check if this is needed:
// Integration tests will be apart
import { mutate } from 'swr'
import { AxiosAPIService } from '../lib/services/axios'

import {
  MOCK_DETAILED_FEEDBACK as data,
  MOCK_USER as user,
} from '../lib/utils/tests-fake-data'

jest.mock('swr', () => ({
  mutate: jest.fn(),
}))

jest.mock('../lib/services/axios', () => ({
  AxiosAPIService: {
    update: jest.fn(),
  },
}))

describe('<AddComment />', () => {
  it('renders with correct UI and sets the proper values when it changes', async () => {
    const { container } = render(<AddComment data={data} user={user} />)

    expect(container).toMatchSnapshot()

    const comment = 'This is a test for Add Comment component'

    const textarea = screen.getByRole('textbox')

    await userEvent.type(textarea, comment)

    expect(textarea.value).toEqual(comment)

    expect(screen.getByText(/characters left/i).textContent).toContain(
      String(250 - comment.length)
    )
  })

  it('renders error message when the form is submitted without content', async () => {
    render(<AddComment data={data} user={user} />)

    const submit = screen.getByRole('button')

    await userEvent.click(submit)

    expect(screen.getByText(/can't be empty!/i)).toBeInTheDocument()
  })
})
