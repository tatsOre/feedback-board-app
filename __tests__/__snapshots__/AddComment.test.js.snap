// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddComment /> renders with correct UI and sets the proper values when it changes 1`] = `
<div>
  <form
    class="bg-white rounded-10 p-6 md:px-8 mt-6 flex flex-wrap justify-between items-center"
  >
    <h2
      class="w-full text-indigo-800 text-lg mb-6"
    >
      Add Comment
    </h2>
    <textarea
      class="w-full text-[13px] md:text-[15px] text-indigo-800 bg-indigo-100 rounded-5 p-4 mb-2 border-indigo-100 hover:border-blue-900 border cursor-pointer"
      maxlength="250"
      placeholder="Type your comment here"
      rows="3"
    />
    <p
      class="text-indigo-500 text-[13px]"
    >
      250
       character
      s
       left
    </p>
    <button
      class="bg-violet-900 hover:bg-violet-500 rounded-10 text-[13px] md:text-sm leading-3 h-10 md:h-11 px-4 md:px-6 min-w-max"
      type="submit"
    >
      Post Comment
    </button>
  </form>
</div>
`;
